docs/architecture/widget_framework.md

```diff
@@ -0,0 +1,90 @@
+# Widget framework
+
+The widget framework comprises common code that provides a "backbone" to our application.
+
+Each screen's content is composed entirely of widgets, and when a screen requests new data, no matter which type it is, the same backend logic is used to look up the screen's configuration, fetch the appropriate real-time data, transform that data into "candidate" widgets, and then assemble some of those candidates together to be sent back to the client and rendered.
+
+## What's a widget?
+
+A widget is a self-contained piece of content. It's essentially a special type of Model (as in MVC) that can be dropped into template slots alongside any number of other widgets to form a full page of content.
+
+Some examples of widgets include the departures list, subway status, alert cards, and image PSAs. However, even basic screen content like the header and footer are widgets as well.
+
+On the backend, each widget implements the [`WidgetInstance` protocol](/lib/screens/v2/widget_instance.ex). The framework calls these protocol functions to find out each widget's priority and where it would like to appear on the screen. The framework then determines how best to position all available widgets within these constraints.
+
+**Widgets do not directly communicate with one another, or even know about the existence of other widgets.** Widgets only communicate with the framework via the aforementioned protocol functions. Once created, a widget has all of the data it needs to render itself.
+
+## Why widgets?
+
+We have many screens, and many different _types_ of screens. Each one needs to display the most rider-relevant info for its specific location.
+
+The widget framework allows us to automate the process of deciding what content is most relevant from moment to moment on a given screen.
+
+The strict isolation of widgets prevents an explosion in complexity and interdependence. It also makes it possible to swap pieces of content out on a screen without worrying about potential side-effects. Business logic (of which there is much since the T has endless edge cases) is corralled inside `WidgetInstance`-implementing modules.
+
+Once a widget is created, it has all of the data it needs. This means all functions on a widget are pure/free of side effects, and easy to test.
+
+See the [`WidgetInstance` protocol](/lib/screens/v2/widget_instance.ex)'s inline documentation for details on each implementation function.
+
+## Request data flow / how the framework logic works
+
+Screens team members can see this as a [flowchart in Miro](https://miro.com/app/board/o9J_lDRaax4=/).
+
+1. Client requests `/v2/api/screen/{id}`. Subsequent steps take place on the server.
```

Comment:
It might be nice if a dev wants to try to "follow along", not so much as a replacement to these steps. You plain English version is much easier to follow than the code by itself.